### PR TITLE
Normalize Gram Newton-Schulz inputs before casting to fp16

### PR DESCRIPTION
**Pull request by vineethsaivs**
`zeropower_via_gram_newtonschulz` cast to the compute dtype before normalizing. fp16 stops at 65504, so a finite fp32 gradient under a loss scale arrived at the iteration as inf, normalized to nan and left the whole matrix zero. A loss scale of 1e4 on gradients of order 1 already reaches that.

Normalize first, then cast: fp16 inputs normalize in fp32 and come back to fp16, bf16 and fp32 normalize as they are. The iteration itself is unchanged and still runs in the compute dtype on a matrix already in [-1, 1].

Test: `DS_ACCELERATOR=cpu python -m pytest tests/unit/v1/ops/muon/test_muon_gram_normalization.py -q`. 12 of 13 fail before, 13 pass after, covering three shapes at scales 1e4 to 1e15 plus two real Muon training steps. The wider muon suite is 113 passed, 29 skipped. yapf and flake8 clean. Apple M2 Pro CPU, fp16 arithmetic selected explicitly.

### CODE REVIEW

**Comment by vineethsaivs** [comment]
Could you approve the CI run? The nine focused CPU tests and changed-file pre-commit checks pass.

File: deepspeed/runtime/zero/muon/original_muon.py [hunk 1]
```diff
@@ -88,13 +88,14 @@ def zeropower_via_gram_newtonschulz(G, steps: int):
...
-    X = G.to(compute_dtype)
+    # Normalize before narrowing precision so a finite gradient or its norm cannot overflow fp16.
+    X = G.float()
     if G.size(-2) > G.size(-1):
         X = X.mT
 
     n, m = X.size(-2), X.size(-1)
 
-    X = X / (X.norm(dim=(-2, -1), keepdim=True) + 1e-7)
+    X = (X / (X.norm(dim=(-2, -1), keepdim=True) + 1e-7)).to(compute_dtype)
```
[the remark this answers is not in the document]
Comment (vineethsaivs) [reply deepspeed/runtime/zero/muon/original_muon.py]: Fixed in b49bb154. Scaling before the norm reduction also handles finite FP32 inputs around `1e30`. All 15 CPU checks pass, including training with a `1e25` loss scale.

**Comment by delock** [comment]
Hi @vineethsaivs , thanks for the catch and the fix.  I think `amax` might be overkill.  convert to fp32 before norm should be safe enough.  Can you check whether the following logic works?  Thanks!
1. if input is BF16, do normalization and the rest in BF16 data type
2. if input is FP16, convert to fp32 before normalization, convert back to fp16 for the rest
3. For FP32 input, do normalization then convert to fp16 for the rest
4. don't do amax before normalize.
5. keep smallest diff if possible.

The logic above should align with most common practice, feel free to discuss further.

**Comment by vineethsaivs** [comment]
Reworked to your logic in 61dbdfd: no amax, fp16 normalizes in fp32 and casts back, bf16 and fp32 normalize as they are, then the cast. Diff is down to one line plus a comment.

One note: without the amax bound the fp32 Frobenius norm still overflows above roughly 1.8e19, since it squares every element. That is far past anything loss scaling reaches, so I dropped the 1e20 and 1e30 cases from the test and it now runs 1e4 to 1e15.

I also left bf16 casting to the gram compute dtype as master does, rather than running the whole iteration in bf16, since that would change today's behaviour. Say the word if you want that too.

**Review by delock — APPROVED** [review]
